### ISSUE TITLE
A couple of minor nitpicks and low hanging fruits

### ISSUE THREAD

**Issue by gkokolatos** [berlin-ab/cexpect#1]
[] It might be useful to try to guard against mutliple header file inclusion. E.g. include the content of each header file within `#ifndef _FILENAME_H_ #define  _FILENAME_H_ #endif` blocks 

[] It is not great (although for example Postgres does it some) use `sizeof(type)` in alloc functions. Especially if that type is `char` then some critique might follow. It is a bit more accepted to write `type *var = allocfunc(nelems * sizeof(*var));` 

[] `exit(func(foo))` is not really great since you probably want to be explicit of the exit status regardless of platform and or shell running it. One of the ways to write this is along the lines of 
``` 
     if (func(foo))
          exit(EXIT_FAILURE);
     else
         exit(EXIT_SUCCESS);
```

[] Also the `main` function in the examples does not have a return value. Although the framework is expected to exit a return value will be asked from most compilers. 

[] Use of `sprintf` is generally discouraged in preference to the safer `snprintf` which takes a size argument.

* In the expression `"%s", (expected_value ? "true" : "false")` the parenthesis means something else, and although most compilers will not complain and do the right thing, they can and should be omitted.

**Comment by jchampio** [comment]
> In the expression `"%s", (expected_value ? "true" : "false")` **the parenthesis means something else**

Can you elaborate? This is new to me.

**Comment by jchampio** [comment]
I'll piggyback on the issue and add a couple more things I noticed here. (I'm happy to split these out into their own issues if that's better.)
- [ ] Contributors need a `LICENSE` to be chosen.
- [ ] If C99 is the development flavor of C in use, that needs to be an explicit choice (best made early on).
- [ ] I'm most used to `struct lower_snake_case` for struct naming in modern C. Things like `Suite *` being passed around add cognitive overhead as I try to figure out whether I'm inside C/C++ hybrid code.
- [ ] Double-newlines as a top-level separator is not as prevalent (or, IMO, as useful) in C as it is in Python, possibly because of the explicit closing braces and because nearly everything is at the top level in C anyway. To my eyes it just makes  everything  look   really    spaced     out. If you use it, use it judiciously.
- [ ] Try to be `const`-correct with pointers, both in your APIs and struct declarations. It's a little bit of a pain at first but it makes things nicer for you and your clients. (`const`-correctness with scalar arguments is IMO not necessary, though you will find people in that camp as well.)
- [x] For declarations like [this](https://github.com/berlin-ab/cexpect/blob/c1962dc35eb647289f2136f461b1571849c11e75/cexpect/cexpect.c#L171), I think you will be happier if you `typedef` your function declarations.

**Comment by berlin-ab** [comment]
@jchampio Can you point me to resources on const-correctness? I'm not familiar with that term.

**Comment by berlin-ab** [comment]
> * It is not great (although for example Postgres does it some) use `sizeof(type)` in alloc functions. Especially if that type is `char` then some critique might follow. It is a bit more accepted to write `type *var = allocfunc(nelems * sizeof(*var));`

I'm trying to understand this one more. Would you elaborate the rationale for this change?

**Comment by gkokolatos** [comment]
> > * It is not great (although for example Postgres does it some) use `sizeof(type)` in alloc functions. Especially if that type is `char` then some critique might follow. It is a bit more accepted to write `type *var = allocfunc(nelems * sizeof(*var));`
> 
> I'm trying to understand this one more. Would you elaborate the rationale for this change?

Of course! Usually two reasons: 1) Easiness of change. If `var` changes type one does not have go all over the code where `var` is referenced to change the alloc sizes appropriately. 2) In the example above`sizeof(*var)` is considered to denote a bit more clearly the intention despite the result being equivalent with `type *var = allocfunc(nelems * sizeof(type));`

As far as the `sizeof(char)` is concerned, in c this is exactly 1 byte, always. I have seen a bit of a critique to the usage in patches in some open source projects (openBSD for example). Personally, I don't mind, yet it is not unheard of attracting critique.

I hope this helps. Also I am by no means a C expert, just conversational :)

**Comment by gkokolatos** [comment]
> > In the expression `"%s", (expected_value ? "true" : "false")` **the parenthesis means something else**
> 
> Can you elaborate? This is new to me.

I guess you are right. I shouldn't be too strict and I should have said, 'possibly mean something else'. Apologies.

The point that the statement was trying to raise was about parenthesized expressions vs ternary expressions. For example `f(foo, (i = 1, i + 1), bar)` the function f takes three arguments, the second one having the value 2. This is what I meant.
Since it is not uncommon to chain ternary expressions together using parenthesized expressions, I was a bit too quick to comment that the parentheses in the example above are not needed or denote chaining which was not the case. My bad.

**Comment by jchampio** [comment]
> Can you point me to resources on const-correctness? I'm not familiar with that term.

@berlin-ab Unfortunately the resources I'm finding are mostly focused on C++ maintenance, like [the Super FAQ](https://isocpp.org/wiki/faq/const-correctness), which means they start pulling in other concepts like const references that aren't very helpful. Many of the same concepts still apply in C, but from an API perspective, you only really have functions and structs to apply them to.

I'll pull concrete examples from [this code](https://github.com/berlin-ab/cexpect/blob/8da0565efe088cc5d5710e46e6b7c2491e238e15/cexpect/cexpect.c#L21):
```c
void add_test_to_suite(Suite *suite, test_function_type test_function, int line_number, char *file_name) {
```
With the above declaration, the compiler allows the function to modify the struct pointed to by `suite`, which callers would probablye expect given the function's apparent goal. But it also allows the function to modify the characters pointed to by `file_name`, in a way that will be visible to the caller. That's potentially dangerous and unexpected (and crash-worthy, if you pass a string constant in over the compiler's complaints).

In this case I would modify the function signature to make it const-correct:
```c
void add_test_to_suite(Suite *suite, test_function_type test_function, int line_number, const char *file_name) {
```
So now `file_name` is a pointer to `const char`, and the API _guarantees_ (via the compiler) that the function cannot make changes through the pointer. (As long as the function implementation doesn't explicitly cast `const` away.) This is how you tell your API consumers "I will not modify the thing you've given me."

You _could_ make `int line_number` const as well, and/or constify the pointers themselves (`const char * const file_name`) but callers won't see changes to those even if you reassign them inside the function, so I don't think there's a reason to be that strict. I also don't think making the function pointer `const` has any effect at all; you shouldn't be able to assign to a dereferenced function pointer anyway.

You can apply similar fixups to the struct definitions as well:
```c
struct FailedTestData {
    char *expected_value;
    char *actual_value;
    int line_number;
    char *file_name;
    void *extra;
};
```
Assuming nothing needs to modify the contents of those strings, they should probably be `const char *`. Maybe you want to make sure that no one can change some of those values for the lifetime of the struct, in which case you could make the members themselves `const`. (That might be overkill for simple APIs, but it's useful for C-style OOP.) And if you wanted to freeze the members of a struct all at once, you could make it a `const struct FailedTestData` when you declare it.

**Comment by berlin-ab** [comment]
Awesome explanation. Thanks @jchampio . I'll give it a shot.
